src/hash_generator/hashAlgorithmUtils.js

```diff
@@ -0,0 +1,209 @@
+ * It useful for the rule n°8 of the algorithm
+ *
+ * @param {string} date - a string representing an UTC date
+ * @return {string} - the date with a 'Z' at the end if it needs to, the date without any
+ * modification otherwise
+ */
+export const addATrailingZIfItIsNeeded = (date) => {
+  if (getTheTimeZoneOffsetFromDateString(date)) return date;
+
+  if (date.endsWith('Z')) return date;
+
+  return `${date}Z`;
+};
+
+/**
+ * Return a date passed in parameter, but with a millisecond property that follows the rule of the
+ * algorithm
+ * @param {string} date
+ * @returns {string} the updated date
+ */
+export const addMillisecondPrecisionToDate = (date) => {
+  const YearMonthDay = date.substring(0, date.indexOf('T'));
+  const time = date.substring(date.indexOf('T') + 1);
+  const startTime = time.substring(0, 5);
+  let endTime = time.substring(6);
+  let ms;
+  let end = getTheTimeZoneOffsetFromDateString(date);
+
+  if (!end) {
+    end = 'Z';
+  }
+
+  endTime = endTime.substring(0, endTime.indexOf(end.charAt(0)));
+
+  if (endTime.length > 2) {
+    ms = endTime.substring(endTime.indexOf('.') + 1);
+    endTime = endTime.substring(0, endTime.indexOf('.'));
+
+    while (ms.length < 3) {
+      ms += '0';
+    }
+
+    if (ms.length > 3) {
+      ms = ms.substring(0, 3);
+    }
+  } else {
+    ms = '000';
+  }
+
+  return `${YearMonthDay}T${startTime}:${endTime}.${ms}${end}`;
+};
+
+/**
+ * If the parameter is a date, it returns the parameter with some modification to follow the
+ * rules defined by the algorithm
+ * If the parameter isn't a date, it returns the parameter
+ *
+ * It useful for the rule n°8 and n°9 of the algorithm
+ *
+ * @param {*} obj
+ * @return {*} - the obj if it isn't a date, the date with some modification to follow the
+ * rules defined by the algorithm otherwise.
+ */
+export const formatTheDateToFollowTheAlgorithmRules = (obj) => {
+  if (!isADate(obj)) return obj;
+
+  let date = addATrailingZIfItIsNeeded(obj);
+
+  date = removeTimeZonePropertyIfItIsNeeded(date);
+
+  date = addMillisecondPrecisionToDate(date);
+
+  return date;
+};
+
+/**
+ * returns a string that concatenated all the string passed in param, ordered in lexical order
+ * @param {[]} strings - the list of string
+ * @returns {string} string that concatenated all the string, ordered in lexical order
+ */
+export const listOfStringToPreHashLexicalOrderedString = (strings) => {
+  let string = '';
+  strings.sort();
+
+  for (let i = 0; i < strings.length; i += 1) {
+    string += strings[i];
+  }
+  return string;
+};
+
+/**
+ * Read and returns all the contexts provided in the object
+ * @param {{}} object
+ * @return {{}} - the contexts
+ * e.g {
+ *    "example": "http://ns.example.com/epcis/",
+ *    "example2": "http://ns.example2.com/epcis/",
+ * }
+ */
+export const readAllTheContextProvidedInTheObject = (object) => {
+  let context = {};
+
+  Object.keys(object).forEach((key) => {
+    if (object[key].toString() === '[object Object]') {
```

Comment:
Could be `typeof object[key] === 'object'`

Also if you need keys and value, try `Object.entries()`